src/modules/adaptors/Convex/ConvexAdaptor.sol

```diff
@@ -0,0 +1,268 @@
+
+        // booster.withdraw(pid, amount);
+
+        IBaseRewardPool baseRewardPool = IBaseRewardPool(rewardPool);
+        baseRewardPool.withdrawAndUnwrap(amount, true); // TODO: not sure if we just always set this to true (might be gas intensive), or if we allow this as a param somehow.
+    }
+
+    /**
+     * @notice Functions Cellars use to determine the withdrawable balance from an adaptor position.
+     * @dev Accounts for LPTs in the Cellar's wallet, and staked in Convex Market.
+     * @param adaptorData see adaptorData info at top of this smart contract
+     * @param configurationData see configurationData at top of this smart contract
+     */
+    function withdrawableFrom(
+        bytes memory adaptorData,
+        bytes memory configurationData
+    ) public view override returns (uint256) {
+        bool isLiquid = abi.decode(configurationData, (bool));
+        if (isLiquid) {
+            (, address rewardPool) = abi.decode(adaptorData, (uint256, address));
+            IBaseRewardPool baseRewardPool = IBaseRewardPool(rewardPool);
+            ERC20 stakingToken = ERC20(baseRewardPool.stakingToken());
+            return (stakingToken.balanceOf(msg.sender) + baseRewardPool.balanceOf(msg.sender));
+        } else return 0;
+    }
+
+    /**
+     * @notice Calculates the Cellar's balance of the positions creditAsset, a specific underlying LPT.
+     * @param adaptorData see adaptorData info at top of this smart contract
+     * @return total balance of LPT for Cellar, including liquid and staked
+     * TODO: This assumes that no rewards are given back as accrual of more curveLPT. I believe that to be the case because BaseRewardPool has its own rewardsToken, and extraRewards has specific reward contracts specific to respective convex markets.
+     */
+    function balanceOf(bytes memory adaptorData) public view override returns (uint256) {
+        (, address rewardPool) = abi.decode(adaptorData, (uint256, address));
+        IBaseRewardPool baseRewardPool = IBaseRewardPool(rewardPool);
+        ERC20 stakingToken = ERC20(baseRewardPool.stakingToken());
+        return (stakingToken.balanceOf(msg.sender) + baseRewardPool.balanceOf(msg.sender));
+    }
+
+    /**
+     * @notice Returns the positions underlying assets.
+     * @param adaptorData see adaptorData info at top of this smart contract
+     * @return Underlying LPT for Cellar's respective Convex market position
+     */
+    function assetOf(bytes memory adaptorData) public view override returns (ERC20) {
+        (uint256 pid, ) = abi.decode(adaptorData, (uint256, address));
+        (address lpToken, , , , , ) = booster.poolInfo(pid);
+        ERC20 lpt = ERC20(lpToken);
+        return lpt;
+        // TODO: decide to use above or alternative way to get lpt below
+        // IBaseRewardPool rewardsPool = IBaseRewardPool(rewardsPool);
```

Comment:
Where does rewards pool come from? I think it is fine to store the Lp address in the adaptor data

Reply:
`rewardsPool` represents the `baseRewardPool` or `crvRewards` object within each `PoolInfo` struct for a Convex Market. The `Booster` contract for Convex-Curve has an array of `PoolInfo` structs corresponding to a `pid` which is just the index of the mentioned array keeping track of all the different Convex Pools (aka Convex Markets). 

So we get `rewardsPool` from the struct by querying the `Booster`. It is queried though via reading multiple storage slots, and I think that is why you pointed out that we should be fine with just the lp address in adaptorData.

I've carried out the check here though but maybe we wanted the check elsewhere (see updated code).